feat: performance tweaks

ojeytonwilliams opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

This should improve the non-image related performance by deferring non-critical things and preloading critical ones. ~~And, as a side benefit, the fonts are never invisible now!~~  This does rely on `rel="preload"` https://caniuse.com/?search=preload 92% usage right now. 

I wasn't 100% sure if we should be using the built files or not.  We seem to have a mixture, which probably isn't right.

<!-- Feel free to add any additional description of changes below this line -->
Edit: the original PR only avoided the FOIT by loading system fonts and never changing.  This version uses `swap` to replace the system fonts asap.

ahmadabdolsaheb commented:
Thanks for your this submission and I am exited to see this functionality in production.

Could you please confirm that you are able to see additional articles with a specific tags loaded when the button is clicked. And that the button disappears when there are no more related articles to load.

Additionally, would it be possible to have this pull request from a separate branch against master.
It would help with managing the flow and the review process.

Thank you

sunnyssr commented:
Yes, it is working fine on my pc. I would send the video within few hours.

ojeytonwilliams committed:
fix: defer more scripts and sheets

Delays execution of several scripts and sheets until the DOM has fully
loaded, which saves a little time and makes Lighthouse happier.

ojeytonwilliams committed:
feat: preload crit assets and swap fonts

ojeytonwilliams committed:
fix: load algolia css normally to avoid FOUC

ojeytonwilliams committed:
fix: add missing 'as' properties

ojeytonwilliams commented:
So, now I understand the built directory. I'll fix this!

raisedadead commented:
Just closed an old issue thread from a couple years ago: https://github.com/freeCodeCamp/freeCodeCamp/issues/36849 - but essentially a good example of moving forward slow and leaving things broken.

ojeytonwilliams commented:
Jan 2019.  Ouch.

ojeytonwilliams committed:
fix: build and clean up

Pointed some asset css/ to asset/built, too

ojeytonwilliams commented:
I also forget to gscan, sorry. I'll put together a basic CI pipeline soon.

ojeytonwilliams committed:
fix: update jQuery, use slim version

ojeytonwilliams committed:
fix: defer fitvids in error pages

scissorsneedfoodtoo commented:
Tested locally with code blocks and embedded videos. Everything LGTM. Seems like these changes will bump the lighthouse score quite a bit.

I did notice a couple of warnings in the dev console. I'll leave suggestions with potential fixes.

scissorsneedfoodtoo commented on default.hbs:
```diff
@@ -12,19 +12,22 @@
+    <noscript><link rel="stylesheet" href="https://cdn.freecodecamp.org/news-assets/prism-1-16-0/prism-unescaped-markup.css"></noscript>
 
-    <link href="https://fonts.googleapis.com/css?family=Lato:300,400,400i,700|Roboto+Mono:400,700" rel="stylesheet">
+    <link rel="preload" as="font" crossorigin href="https://fonts.googleapis.com/css?family=Lato:300,400,400i,700|Roboto+Mono:400,700" rel="stylesheet">
```
This seems to get rid of the `The resource ... was preloaded using link preload but not used within a few seconds from the window's load event` warning:

```suggestion
    <link rel="preload stylesheet" as="style" crossorigin href="https://fonts.googleapis.com/css?family=Lato:300,400,400i,700|Roboto+Mono:400,700" />
```

ojeytonwilliams replied:
Oh, wow, I had two `rel`s :facepalm:

I *think* we need to just use "preload", though, like this https://stackoverflow.com/questions/61889044/how-to-use-both-rel-preload-and-rel-stylesheet-for-the-same-tag suggests.  Also, I'll need to check on Firefox to see if the preload issue has been solved.

scissorsneedfoodtoo replied:
Awesome, thanks for catching that. I didn't think to check for browser differences with preload. Yeah, let's try to figure out one of the methods in that thread that works with all browsers and possibly a fallback.

scissorsneedfoodtoo commented on default.hbs:
```diff
@@ -12,19 +12,22 @@
     {{!-- Styles'n'Scripts --}}
     <link rel="stylesheet" type="text/css" href="{{asset "built/screen.css"}}" />
-    <link rel="stylesheet" type="text/css" href="{{asset "css/algolia-search.css"}}" />
+    <link rel="preload" as="style" href="{{asset "built/global.css"}}" />
```
This also seems to get rid of the same warning:

```suggestion
    <link rel="preload stylesheet" as="style" href="{{asset "built/global.css"}}" />
```

ojeytonwilliams committed:
fix: one onload, one rel

Using onload instead of using both a preload and a stylesheet link means
that only one resource gets downloaded.

ojeytonwilliams commented:
@scissorsneedfoodtoo I think I've finally got the right attributes.  Could you check if the console warnings have gone?

Re: firefox, it's working perfectly locally.

scissorsneedfoodtoo commented:
@ojeytonwilliams, just tested in Chrome and Firefox and am seeing nice improvements in the lighthouse score. Also, I no longer see those warnings in the console.

The only thing I noticed is that when I search for something on the search results page it says `index` is undefined. I'll play around with it and see if I can make a suggestion somewhere.

scissorsneedfoodtoo commented on search-results.hbs:
```diff
@@ -10,6 +10,7 @@
 </main>
 
 <script>
+document.addEventListener('DOMContentLoaded', () => {
```
It's not the DRYest, but this seems to fix the issue:

```suggestion
document.addEventListener('DOMContentLoaded', () => {
    const client = algoliasearch('QMJYL5WYTI', '89770b24481654192d7a5c402c6ad9a0');
    const index = client.initIndex(main['algolia-index-name']);
```

There may be a way to set this up globally, but it may be tricky with the deferred Algolia scripts and the `DOMContentLoaded` events.

ojeytonwilliams replied:
This seems good.  I do get a few errors and a 404 on the /search page, for both this and `main`.  Is there a trick to getting that working locally?

ojeytonwilliams committed:
Update search-results.hbs

Co-authored-by: Kristofer Koishigawa <scissorsneedfoodtoo@gmail.com>

scissorsneedfoodtoo committed:
fix: algolia-search and search-results spacing

scissorsneedfoodtoo committed:
fix: submitSearch not defined error

raisedadead commented:
Hey @scissorsneedfoodtoo please deploy.

scissorsneedfoodtoo commented:
Just deployed to all the News instances. Thanks again @ojeytonwilliams for all of these enhancements.
